### PR TITLE
better msgpack decoding

### PR DESCRIPTION
**Pull request by gnvk**

### CODE REVIEW

**Review by leki75 — APPROVED** [review]
Nice, nice, nice!

**Comment by gnvk** [comment]
NOTE: this can be even further improved by not calling msgpack.Unmarshal but do it in the loop by hand.

File: v2/stream/datav2.go [hunk 1]
```diff
@@ -258,47 +258,94 @@ func (s *datav2stream) readForever() {
...
 
+	return nil
+}
+
+func (s *datav2stream) handleMessage(T string, msg []byte) error {
 	switch T {
 	case "t", "q", "b":
 	default:
 		return nil
 	}
```
Comment (leki75) [inline v2/stream/datav2.go:338]: This is not needed anymore.

**Comment by gnvk** [comment]
Benchmark results:

- msgpack unmarshal to map, mapstructure decode
```
BenchmarkHandleMessages
BenchmarkHandleMessages-12    	   38446	     30357 ns/op	   15998 B/op	     242 allocs/op
PASS
ok  	github.com/alpacahq/alpaca-trade-api-go/v2/stream	1.496s
```

- msgpack unmarshal for T, msgpack unmarshal for the rest
```
BenchmarkHandleMessages
BenchmarkHandleMessages-12    	  178768	      6327 ns/op	     596 B/op	      16 allocs/op
PASS
ok  	github.com/alpacahq/alpaca-trade-api-go/v2/stream	1.211s
```

- msgpack custom unmarsal for the whole message
```
BenchmarkHandleMessages
BenchmarkHandleMessages-12    	  446404	      2351 ns/op	     104 B/op	      12 allocs/op
PASS
ok  	github.com/alpacahq/alpaca-trade-api-go/v2/stream	1.088s
```

File: v2/stream/datav2.go [hunk 2]
```diff
@@ -258,82 +258,209 @@ func (s *datav2stream) readForever() {
...
+
+		switch T {
+		case "t":
+			err = s.handleTrade(d, n)
+		case "q":
+			err = s.handleQuote(d, n)
+		case "b":
+			err = s.handleBar(d, n)
+		default:
+			// TODO!
```
Comment (kalmant) [inline v2/stream/datav2.go:310]: TODO left in

File: v2/stream/entities.go [hunk 3]
```diff
@@ -4,37 +4,37 @@ import "time"
...
-	Close     float64   `mapstructure:"c"`
-	Volume    uint64    `mapstructure:"v"`
-	Timestamp time.Time `mapstructure:"t"`
+	Symbol    string    `msgpack:"S"`
+	Open      float64   `msgpack:"o"`
+	High      float64   `msgpack:"h"`
+	Low       float64   `msgpack:"l"`
+	Close     float64   `msgpack:"c"`
+	Volume    uint64    `msgpack:"v"`
+	Timestamp time.Time `msgpack:"t"`
```
Comment (kalmant) [inline v2/stream/entities.go:39]: Why do these fields have tags?

File: v2/stream/datav2_test.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+	Symbol:    "TEST",
+	Open:      100,
+	High:      101.2,
+	Low:       98.67,
+	Close:     101.1,
+	Volume:    2560,
+	Timestamp: time.Date(2021, 03, 05, 16, 0, 0, 0, time.UTC),
+}
+
+func TestHandleMessages(t *testing.T) {
```
Comment (kalmant) [inline v2/stream/datav2_test.go:91]: What about adding some unexpected fields to these structs? That would help ensure the API can be extended later on without breaking it for the users of the go API.

**Review by kalmant — APPROVED** [review]
🆒

**Pull request merged**